File: agent/tools/ValidateTool.js
const fs = require('fs');
const { exec } = require('child_process');
const BaseTool = require('./BaseTool');

class ValidateTool extends BaseTool {
  async execute() {
    this.log('✅ Validating i18n implementation...');
    
    // Get transform results from agent state
    const transformResults = this.agent.transformResults || { files: {} };
    
    const results = {
      validFiles: 0,
      invalidFiles: 0,
      files: {},
      issues: [],
      retryAttempts: 0,
      maxRetries: 3
    };
    
    // Validate each file with retry mechanism
    for (const [filePath, fileResult] of Object.entries(transformResults.files)) {
      if (fileResult.success) {
        const validation = await this.validateFileWithRetry(filePath, results.maxRetries);
        results.files[filePath] = validation;
        
        if (validation.isValid) {
          results.validFiles++;
        } else {
          results.invalidFiles++;
          results.issues.push(...validation.issues);
        }
      }
    }
    
    // Store results in agent state
    this.agent.validateResults = results;
    
    return results;
  }
  
  async validateFileWithRetry(filePath, maxRetries) {
    let lastValidation = null;
    
    for (let attempt = 1; attempt <= maxRetries; attempt++) {
      this.log(`  Validating ${filePath} (attempt ${attempt}/${maxRetries})...`);
      
      const validation = await this.validateFile(filePath);
      lastValidation = validation;
      
      if (validation.isValid) {
        this.success(`File ${filePath} is valid`);
        return validation;
      } else {
        this.error(`File ${filePath} has issues: ${validation.issues.join(', ')}`);
        
        if (attempt < maxRetries) {
          this.log(`  Attempting to fix issues in ${filePath}...`);
          const fixed = await this.attemptFix(filePath, validation.issues);
          if (fixed) {
            this.log(`  Fixed issues in ${filePath}, re-validating...`);
            continue;
          } else {
            this.error(`  Could not fix issues in ${filePath}`);
          }
        }
      }
    }
    
    // If we get here, all retries failed
    this.error(`File ${filePath} failed validation after ${maxRetries} attempts`);
    return {
      ...lastValidation,
      isValid: false,
      issues: [...(lastValidation?.issues || []), `Failed after ${maxRetries} retry attempts`]
    };
  }
  
  async validateFile(filePath) {
    if (!fs.existsSync(filePath)) {
      return { isValid: false, issues: ['File not found'] };
    }
    
    const content = fs.readFileSync(filePath, 'utf8');
    const issues = [];
    
    // 1. Check for syntax errors first (most critical)
    const syntaxErrors = await this.checkSyntaxErrors(filePath, content);
    if (syntaxErrors.length > 0) {
      issues.push(...syntaxErrors);
      return { isValid: false, issues: issues, critical: true };
    }
    
    // 2. Check for useTranslation import
    if (!content.includes('useTranslation')) {
      issues.push('Missing useTranslation import');
    }
    
    // 3. Check for t() function usage
    if (!content.includes('t(')) {
      issues.push('No t() function calls found');
    }
    
    // 4. Check for hardcoded strings that should be translated
    const hardcodedStrings = this.findRemainingHardcodedStrings(content);
    if (hardcodedStrings.length > 0) {
      issues.push(`Found ${hardcodedStrings.length} remaining hardcoded strings`);
    }
    
    // 5. Check for proper i18n setup
    const i18nIssues = this.checkI18nSetup(content);
    if (i18nIssues.length > 0) {
      issues.push(...i18nIssues);
    }
    
    return {
      isValid: issues.length === 0,
      issues: issues,
      hardcodedStrings: hardcodedStrings,
      critical: false
    };
  }
  
  findRemainingHardcodedStrings(content) {
    const strings = [];
    const lines = content.split('\n');
    
    // Look for hardcoded strings that should have been translated
    const patterns = [
      />\s*([A-Z][^<{]+?)\s*</g, // JSX text content
      /notify\(['"]([^'"]+)['"]/g, // Notify messages
      /<Button[^>]*>([^<]+)<\/Button>/g, // Button text
    ];
    
    lines.forEach((line, lineIndex) => {
      patterns.forEach(pattern => {
        let match;
        while ((match = pattern.exec(line)) !== null) {
          const text = match[1].trim();
          if (this.isLikelyTranslatable(text)) {
            strings.push({
              text: text,
              line: lineIndex + 1,
              context: line.trim()
            });
          }
        }
      });
    });
    
    return strings;
  }
  
  isLikelyTranslatable(text) {
    // Check if text looks like it should be translated
    if (text.length < 3) return false;
    if (/^[A-Z_]+$/.test(text)) return false; // Constants
    if (text.includes('${') || text.includes('{{')) return false; // Template literals
    if (text.startsWith('/')) return false; // Routes
    
    // Check if it's a common UI string
    const commonUIStrings = [
      'Save', 'Edit', 'Delete', 'Add', 'Remove', 'Update', 'Cancel', 'Submit',
      'Login', 'Logout', 'Register', 'Sign up', 'Sign in', 'Welcome', 'Hello',
      'Error', 'Success', 'Warning', 'Info', 'Loading', 'Please wait'
    ];
    
    return commonUIStrings.some(uiString => 
      text.toLowerCase().includes(uiString.toLowerCase())
    );
  }
  
  async checkSyntaxErrors(filePath, content) {
    const issues = [];
    
    // 1. Check for basic syntax errors using Node.js
    try {
      // Try to parse the file as JavaScript
      const vm = require('vm');
      const context = { require, module, exports, __dirname, __filename };
      vm.createContext(context);
      
      // Remove JSX and try to parse as JS
      const jsContent = content
        .replace(/import.*from.*['"].*['"];?\n?/g, '') // Remove imports
        .replace(/export default.*;?\n?/g, '') // Remove exports
        .replace(/<[^>]*>/g, 'null') // Replace JSX with null
        .replace(/jsx`/g, '`') // Fix template literals
        .replace(/`jsx/g, '`'); // Fix template literals
      
      vm.runInContext(jsContent, context);
    } catch (error) {
      issues.push(`Syntax error: ${error.message}`);
    }
    
    // 2. Check for common i18n syntax errors
    if (content.includes('t(\'\')') || content.includes('t("")')) {
      issues.push('Empty t() function call found');
    }
    
    // Check for unclosed t() calls
    const unclosedT = content.match(/t\(['"][^'"]*$/g);
    if (unclosedT) {
      issues.push('Unclosed t() function call found');
    }
    
    // 3. Check for proper useTranslation usage
    if (content.includes('useTranslation') && !content.includes('const { t }')) {
      issues.push('useTranslation imported but not destructured');
    }
    
    // 4. Check for missing semicolons or brackets
    const lines = content.split('\n');
    let bracketCount = 0;
    let parenCount = 0;
    
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      
      // Count brackets and parentheses
      for (const char of line) {
        if (char === '{') bracketCount++;
        if (char === '}') bracketCount--;
        if (char === '(') parenCount++;
        if (char === ')') parenCount--;
      }
    }
    
    if (bracketCount !== 0) {
      issues.push(`Unmatched brackets: ${bracketCount > 0 ? 'missing' : 'extra'} ${Math.abs(bracketCount)} closing bracket(s)`);
    }
    
    if (parenCount !== 0) {
      issues.push(`Unmatched parentheses: ${parenCount > 0 ? 'missing' : 'extra'} ${Math.abs(parenCount)} closing parenthesis`);
    }
    
    return issues;
  }
  
  checkI18nSetup(content) {
    const issues = [];
    
    // Check if useTranslation is properly imported
    if (content.includes('useTranslation')) {
      if (!content.includes('from \'react-i18next\'') && !content.includes('from "react-i18next"')) {
        issues.push('useTranslation imported from wrong package');
      }
      
      if (!content.includes('const { t } = useTranslation()')) {
        issues.push('useTranslation not properly destructured');
      }
    }
    
    // Check for proper t() usage patterns
    const tCalls = content.match(/t\(['"][^'"]*['"]\)/g) || [];
    if (tCalls.length > 0) {
      // Check for proper key naming
      tCalls.forEach(call => {
        const key = call.match(/t\(['"]([^'"]*)['"]\)/)[1];
        if (key.length < 2) {
          issues.push(`Invalid translation key: "${key}" (too short)`);
        }
        if (!key.includes('.') && key.length > 10) {
          issues.push(`Translation key should use dot notation: "${key}"`);
        }
      });
    }
    
    return issues;
  }
  
  async attemptFix(filePath, issues) {
    try {
      const content = fs.readFileSync(filePath, 'utf8');
      let fixedContent = content;
      let fixed = false;
      
      // Fix common issues
      for (const issue of issues) {
        if (issue.includes('Missing useTranslation import')) {
          // Add useTranslation import
          if (!fixedContent.includes('import { useTranslation }')) {
            const importLine = "import { useTranslation } from 'react-i18next';\n";
            fixedContent = importLine + fixedContent;
            fixed = true;
          }
        }
        
        if (issue.includes('useTranslation imported but not destructured')) {
          // Add destructuring
          if (fixedContent.includes('useTranslation') && !fixedContent.includes('const { t }')) {
            // Find the function component and add destructuring
            const functionMatch = fixedContent.match(/function\s+\w+\s*\([^)]*\)\s*{/);
            if (functionMatch) {
              const insertPoint = functionMatch.index + functionMatch[0].length;
              fixedContent = fixedContent.slice(0, insertPoint) + 
                '\n  const { t } = useTranslation();\n' + 
                fixedContent.slice(insertPoint);
              fixed = true;
            }
          }
        }
        
        if (issue.includes('Unclosed t() function call')) {
          // Try to fix unclosed t() calls
          fixedContent = fixedContent.replace(/t\(['"][^'"]*$/gm, (match) => {
            return match + '")';
          });
          fixed = true;
        }
      }
      
      if (fixed) {
        fs.writeFileSync(filePath, fixedContent);
        return true;
      }
      
      return false;
    } catch (error) {
      this.error(`Failed to fix ${filePath}: ${error.message}`);
      return false;
    }
  }
}

module.exports = ValidateTool;

File: pythonAgent/tools/validate_tool.py
"""
Validate tool for checking i18n implementation.
"""

import re
import subprocess
from pathlib import Path
from typing import Dict, List, Any
from tools.base_tool import BaseTool


class ValidateTool(BaseTool):
    """Tool for validating i18n implementation."""
    
    async def execute(self) -> dict:
        """Execute the validate tool."""
        self.log('  Validating i18n implementation...')
        
        # Get transform results from agent state
        state = self.agent.state_manager.get_state()
        transform_results = state.transform_results or {'files': {}}
        
        results = {
            'valid_files': 0,
            'invalid_files': 0,
            'files': {},
            'issues': [],
            'retry_attempts': 0,
            'max_retries': 3
        }
        
        # Validate each file with retry mechanism
        for file_path, file_result in transform_results.get('files', {}).items():
            if file_result.get('success'):
                validation = await self._validate_file_with_retry(file_path, results['max_retries'])
                results['files'][file_path] = validation
                
                if validation['is_valid']:
                    results['valid_files'] += 1
                else:
                    results['invalid_files'] += 1
                    results['issues'].extend(validation['issues'])
        
        # Store results in agent state
        self.agent.state_manager.update_state({'validate_results': results})
        
        # Update phase to indicate validation is complete
        if results['invalid_files'] == 0:
            self.agent.state_manager.update_state({'phase': 'validation_complete'})
            self.log(f'    Validation phase complete - {results["valid_files"]} files valid, {results["invalid_files"]} files invalid')
        else:
            self.log(f'    Validation phase complete - {results["valid_files"]} files valid, {results["invalid_files"]} files invalid with issues')
        
        return results
    
    async def _validate_file_with_retry(self, file_path: str, max_retries: int) -> Dict[str, Any]:
        """Validate a file with retry mechanism."""
        last_validation = None
        
        for attempt in range(1, max_retries + 1):
            self.log(f'    Validating {file_path} (attempt {attempt}/{max_retries})...')
            
            validation = await self._validate_file(file_path)
            last_validation = validation
            
            if validation['is_valid']:
                self.log(f'    File {file_path} is valid')
                return validation
            else:
                self.error(f'    File {file_path} has issues: {", ".join(validation["issues"])}')
                
                if attempt < max_retries:
                    self.log(f'    Attempting to fix issues in {file_path}...')
                    fixed = await self._attempt_fix(file_path, validation['issues'])
                    if fixed:
                        self.log(f'    Fixed issues in {file_path}, re-validating...')
                        continue
                    else:
                        self.error(f'    Could not fix issues in {file_path}')
        
        # If we get here, all retries failed
        self.error(f'    File {file_path} failed validation after {max_retries} attempts')
        return {
            **last_validation,
            'is_valid': False,
            'issues': last_validation['issues'] + [f'Failed after {max_retries} retry attempts']
        }
    
    async def _validate_file(self, file_path: str) -> Dict[str, Any]:
        """Validate a single file."""
        file_path_obj = Path(file_path)
        if not file_path_obj.exists():
            return {'is_valid': False, 'issues': ['File not found']}
        
        content = file_path_obj.read_text(encoding='utf-8')
        issues = []
        
        # 1. Check for syntax errors first (most critical)
        syntax_errors = await self._check_syntax_errors(file_path, content)
        if syntax_errors:
            issues.extend(syntax_errors)
            return {'is_valid': False, 'issues': issues, 'critical': True}
        
        # 2. Check for useTranslation import
        if 'useTranslation' not in content:
            issues.append('Missing useTranslation import')
        
        # 3. Check for t() function usage
        if 't(' not in content:
            issues.append('No t() function calls found')
        
        # 4. Check for hardcoded strings that should be translated
        hardcoded_strings = self._find_remaining_hardcoded_strings(content)
        if hardcoded_strings:
            issues.append(f'Found {len(hardcoded_strings)} remaining hardcoded strings')
        
        # 5. Check for proper i18n setup
        i18n_issues = self._check_i18n_setup(content)
        if i18n_issues:
            issues.extend(i18n_issues)
        
        return {
            'is_valid': len(issues) == 0,
            'issues': issues,
            'hardcoded_strings': hardcoded_strings,
            'critical': False
        }
    
    async def _check_syntax_errors(self, file_path: str, content: str) -> List[str]:
        """Check for syntax errors in the file."""
        try:
            # Try to run a basic syntax check
            if file_path.endswith('.jsx') or file_path.endswith('.js'):
                # For JSX/JS files, we can't easily check syntax without a proper parser
                # Just do basic checks
                if content.count('{') != content.count('}'):
                    return ['Mismatched braces']
                if content.count('(') != content.count(')'):
                    return ['Mismatched parentheses']
                if content.count('[') != content.count(']'):
                    return ['Mismatched brackets']
            return []
        except Exception as error:
            return [f'Syntax check failed: {error}']
    
    def _find_remaining_hardcoded_strings(self, content: str) -> List[Dict[str, Any]]:
        """Find hardcoded strings that should have been translated."""
        strings = []
        lines = content.split('\n')
        
        # Look for hardcoded strings that should have been translated
        patterns = [
            re.compile(r'>\s*([A-Z][^<{]+?)\s*<'),  # JSX text content
            re.compile(r'notify\([\'"]([^\'"]+)[\'"]'),  # Notify messages
            re.compile(r'<Button[^>]*>([^<]+)</Button>'),  # Button text
        ]
        
        for line_index, line in enumerate(lines):
            for pattern in patterns:
                matches = pattern.findall(line)
                for match in matches:
                    text = match.strip()
                    if self._is_likely_translatable(text):
                        strings.append({
                            'text': text,
                            'line': line_index + 1,
                            'context': line.strip()
                        })
        
        return strings
    
    def _is_likely_translatable(self, text: str) -> bool:
        """Check if a string is likely to be user-facing and translatable."""
        if len(text) < 2:
            return False
        
        # Skip technical strings
        technical_patterns = [
            r'^[a-z_]+$',  # snake_case
            r'^[A-Z][a-z]+[A-Z]',  # camelCase
            r'^[a-z]+-[a-z]+',  # kebab-case
            r'^\d+$',  # numbers only
            r'^[a-z]+://',  # URLs
            r'^#',  # CSS selectors
            r'^\.',  # CSS classes
        ]
        
        for pattern in technical_patterns:
            if re.match(pattern, text):
                return False
        
        # Must contain at least one letter
        if not re.search(r'[a-zA-Z]', text):
            return False
        
        return True
    
    def _check_i18n_setup(self, content: str) -> List[str]:
        """Check for proper i18n setup in the file."""
        issues = []
        
        # Check if file has both useTranslation import and t() usage
        has_import = 'useTranslation' in content
        has_usage = 't(' in content
        
        if has_usage and not has_import:
            issues.append('Using t() function but missing useTranslation import')
        
        if has_import and not has_usage:
            issues.append('Importing useTranslation but not using t() function')
        
        return issues
    
    async def _attempt_fix(self, file_path: str, issues: List[str]) -> bool:
        """Attempt to fix issues in a file."""
        # For now, just return False - we don't have auto-fix logic
        # This could be enhanced to automatically fix common issues
        return False
